File: ik_problem/strings_arrays/zero_sum.rb
def zero_subarray(arr)
  dict = {}
  sum = 0
  i = 0
  result = []

  while i < arr.length
    sum += arr[i]

    if dict.keys.include?(sum)
      sub_array = arr[dict[sum]+1..i]
      result.push(sub_array)

      if sub_array.size > 2 && sub_array.include?(0)
        result.push(sub_array[sub_array.index(0)+1..-1])
      end
    else
      dict[sum] = i
    end

    i += 1
  end

  output = ""

  if result
    for sub in result do
      output += sub.join(", ") + "\n"
    end
  end

  return output
end


arr_1 = [6, 0, 1, 2, 3, 4, -10]
arr_2 = [1, 0]
arr_3 = [4, 0, 1, 2, -3]
arr_4 = [3, 1, 2, 3]

puts zero_subarray(arr_1)
puts ""
puts zero_subarray(arr_2)
puts ""
puts zero_subarray(arr_3)
puts ""
puts zero_subarray(arr_4)

File: ik_problems/strings_arrays/zero_sum.py
def zero_sum(a_list):
    h = {}
    total = 0
    res = []
    i = 0

    while i < len(a_list):
        total += a_list[i]

        if total in h.keys():
            sub_set = a_list[h[total]+1:i+1]
            res.append(sub_set)
            if len(sub_set) > 2 and 0 in sub_set:
                res.append(sub_set[sub_set.index(0)+1:])
        else:
            h[total] = i

        i += 1

    if res:
        for sub in res:
            print sub



arr_1 = [6, 0, 1, 2, 3, 4, -10]
arr_2 = [1, 0]
arr_3 = [4, 0, 1, 2, -3]
arr_4 = [3, 1, 2, 3]

zero_sum(arr_1)
zero_sum(arr_2)
zero_sum(arr_3)
zero_sum(arr_4)
